File: book-weight-calculate.ts
// Constant of how heavy a page is
const gramsPerSquareMM = 0.0000801667;

// Function to ask the user for  a number
function promptContinuouslyForValidPositiveInt(prompt: string): number {
  // Loop until we get a valid input
  let success = false;
  let value = 0;

  while (!success) {
    // Prompt the user to enter a number
    const input = window.prompt(prompt);
    // Validate that they entered something
    if (!input) {
      console.warn("Enter a number");
      continue;
    }
    // Try/catch in case parsing the number fails
    try {
      const res = Number.parseInt(input);
      // Make sure it's a real number
      if (!Number.isNaN(res)) {
        if (res < 0) {
          console.warn("Number greater than zero please");
          continue;
        }
        // Make the number of pages available to other code outside the loop
        value = res;
        // Exit the loop
        success = true;
      } else console.warn("Number please");
    } catch {
      console.warn("Enter a valid number please");
    }
  }
  return value;
}

// Get page dimensions
const pageWidth = promptContinuouslyForValidPositiveInt(
  "How wide is your book in mm?"
);
const pageHeight = promptContinuouslyForValidPositiveInt(
  "How tall is your book in mm?"
);

// Get the page count
const pageCount = promptContinuouslyForValidPositiveInt(
  "How many pages in your book?"
);

// Calculate the area of the page in mm²
const pageArea = pageWidth * pageHeight;

// Calculate the weight of a single page
const pageWeight = pageArea * gramsPerSquareMM;

// Calculate the book weight
const weight = pageCount * pageWeight;

// Output the weight to the user
console.log(`You book weighs ${weight / 1000}kg`);

File: book-weight-calculate.py
# Constant of how heavy a page is
gramsPerSquareMM = 0.0000801667

# Function to ask the user for  a number


def promptContinuouslyForValidInt(
    prompt: str,
    maxValue: int = 15000,
    minValue: int = 1
) -> int:
    # Loop until we get a valid input
    success = False
    value = 0

    while (not success):
        # Prompt the user to enter a number
        inputtedValue = input(prompt+" ")
        # Validate that they entered something
        if len(inputtedValue) == 0:
            print("Enter a number")
            continue

        # Try/catch in case parsing the number fails
        try:
            res = int(inputtedValue)
            # Make sure it's a real number
            if (res < minValue):
                print(f"Number greater than {minValue} please")
                continue

            if (res > maxValue):
                print(f"Number less than {maxValue} please")
                continue

            # Make the number of pages available to other code outside the loop
            value = res
            # Exit the loop
            success = True
        except Exception as err:
            print(f"Something went wrong parsing your number: {err}")

    return value


# Get page dimensions
pageWidth = promptContinuouslyForValidInt(
    "How wide is your book in mm?",
    maxValue=1000
)
pageHeight = promptContinuouslyForValidInt(
    "How tall is your book in mm?",
    maxValue=1000
)

# Get the page count
pageCount = promptContinuouslyForValidInt("How many pages in your book?")

# Calculate the area of the page in mm²
pageArea = pageWidth * pageHeight

# Calculate the weight of a single page
pageWeight = pageArea * gramsPerSquareMM

# Calculate the book weight
weight = pageCount * pageWeight

# Output the weight to the user
print(f"You book weighs {weight / 1000}kg")
